### DISCUSSION TITLE
npx prisma init tries to downlaod prisma engines from Node-API for debian-openssl-1.1.x but inside company network

### DISCUSSION THREAD

**Discussion by alexmondaini** [prisma/prisma#24140] [Q&A]
### Question

I am inside a server on my company network and I dont have access to this endpoint where prisma binaries are hosted. I can only download them from my windows laptop , not the server i am trying to run prisma. What are my options in this case ?

How can I make prisma to work ?

the error is : ENOTFOUND binaries.prisma.sh

### How to reproduce (optional)

npx install prisma --save-dev

npx prisma init

prisma version 5.13

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 20.10.0

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @alexmondaini 👋 

One possible solution is to manually download the necessary Prisma binaries on a machine that has access to the internet (like your Windows laptop), and then transfer them to your server. You can then set environment variables to point Prisma to the location of these binaries on your server. See the using custom engine libraries section in the [documentation](https://www.prisma.io/docs/orm/more/under-the-hood/engines#using-custom-engine-libraries-or-binaries).

**Comment by alexmondaini** [comment]
Hi @ludralph,

I've managed to make it work yesterday after some painful attempts 👍. I've transferred the query engine library and the schema engine binary files and it seems to work fine so far.

Is that enough for all scenarios those files?

I see that there is also a binary for the query engine alongside the library. What is the difference between them?

The migration and introspection engine I did not download but it is working fine

**Reply by ludralph** [reply]
> Is that enough for all scenarios those files?

Yes it is.

> I see that there is also a binary for the query engine alongside the library. What is the difference between them?

The Prisma Query Engine can be configured to run in two modes: as a Node-API library or as a binary.

- Library: In this mode, the Query Engine is loaded as a Node-API library into Prisma Client. This is the default and recommended approach as it reduces the communication overhead between the Prisma Client and the Query Engine. This is because the Query Engine runs in the same process as your application, allowing for direct communication.

- Binary: In this mode, the Query Engine is compiled as an executable binary that runs in its own process. The Prisma Client communicates with this process to execute queries. This was the only available mode until Prisma 3.x and is still available as an option.

You can configure the query engine type by providing the engineType property to the Prisma Client generator in your Prisma schema file:
```prisma
generator client {  
 provider = "prisma-client-js"  
 engineType = "binary"  
}
```
Valid values for engineType are `binary` and `library`. You can also use the environment variable `PRISMA_CLIENT_ENGINE_TYPE` to set the engine type.

See the [documentation](https://www.prisma.io/docs/orm/more/under-the-hood/engines#defining-the-query-engine-type-for-prisma-client) for more information.

**Reply by janpio** [reply]
Please don't use the `binary` query engine if there is no good reason for that @alexmondaini. It really is just there for legacy reasons, the Node-API library should work better in almost all cases.

**Comment by alexmondaini** [comment]
Hi @ludralph ,

Is there any other way to get the prisma engine besides this approach ? my company now has blocked the cloudfare endpoint for downloading prisma engines, so I will not be able to update that in the future, which is just sa headache in the longer term

**Reply by ludralph** [reply]
Yes, there are a few alternatives to manually downloading the Prisma engine if your company has blocked the Cloudflare endpoint.

- Using a VPN or different network: If the issue is due to network restrictions, you might be able to bypass them by using a VPN or a different internet connection.

- Building the engines from source: You can build the Prisma engines from the source code. This involves cloning the Prisma engines repository and building the binaries. Once built, you can consume your engines build. Here are the links to the instructions: [Building Prisma Engines](https://github.com/prisma/prisma-engines#building-prisma-engines) and [Using Custom Engine Libraries or Binaries](https://www.prisma.io/docs/concepts/components/prisma-engines#using-custom-engine-libraries-or-binaries).

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
